src/iterator.rs

```diff
@@ -47,10 +48,16 @@ impl GoIter {
 }
 
 impl StorageIterator for GoIter {
-    fn next(&mut self) -> FfiResult<NextItem> {
+    fn next(&mut self) -> FfiResult<Option<KV>> {
         let next_db = match self.vtable.next_db {
             Some(f) => f,
-            None => return Err(FfiError::other("iterator vtable not set")),
+            None => {
+                return (
+                    Err(FfiError::unknown("iterator vtable not set")),
+                    // TODO: should this remain free?
+                    GasInfo::with_externally_used(0),
+                );
```

Comment:
`FfiError::unknown` halts the contract so there's no reason it should really charge gas in this case, imo

Reply:
The gas metering has two purposes:
1. it limits the cost of a transaction
2. it charges a price depending on the computational cost that really occurred

Point 2. is handled relatively well in Ethereum, but not so much in Cosmos. Here you pay the entire block space you allocated. However, this can change at any point (https://github.com/cosmos/cosmos-sdk/issues/2150). So I think we should report a proper gas usage for failing transactions as well.

Reply:
Uhm, this returning zero just means we don't add more to the gas that has already been charged. all the gas that was charged before this point will still be reported back.

Reply:
Ah, okay, so this is more about the base cost of a `GoIter::next` call that is charged no matter if we call info Go land or not. What about adding a little constant here, like 5 or 10?

Reply:
if it's below 100 it won't even be registered outside the wasm module, so i guess you meant 500-1000?
in any case, I'd say it's not really necessary - this branch is a vm-bug branch anyways so you'll be penalizing users for our bugs.

Reply:
It is free now